compiler/rustc_parse/src/parser/diagnostics.rs

```diff
@@ -1076,6 +1120,145 @@ impl<'a> Parser<'a> {
         Ok(())
     }
 
+    pub(super) fn maybe_recover_from_prefix_increment(
+        &mut self,
+        operand_expr: P<Expr>,
+        op_span: Span,
+        prev_is_semi: bool,
+    ) -> PResult<'a, P<Expr>> {
+        let kind = IncDecRecovery {
+            standalone: prev_is_semi,
+            op: IncOrDec::Inc,
+            fixity: UnaryFixity::Pre,
+        };
+
+        self.recover_from_inc_dec(operand_expr, kind, op_span)
+    }
+
+    pub(super) fn maybe_recover_from_postfix_increment(
+        &mut self,
+        operand_expr: P<Expr>,
+        op_span: Span,
+        prev_is_semi: bool,
+    ) -> PResult<'a, P<Expr>> {
+        let kind = IncDecRecovery {
+            standalone: prev_is_semi,
+            op: IncOrDec::Inc,
+            fixity: UnaryFixity::Post,
+        };
+
+        self.recover_from_inc_dec(operand_expr, kind, op_span)
+    }
+
+    fn recover_from_inc_dec(
+        &mut self,
+        base: P<Expr>,
+        kind: IncDecRecovery,
+        op_span: Span,
+    ) -> PResult<'a, P<Expr>> {
+        let mut err = self.struct_span_err(
+            op_span,
+            &format!("Rust has no {} {} operator", kind.fixity, kind.op.name()),
+        );
+        err.span_label(op_span, &format!("not a valid {} operator", kind.fixity));
+
+        if let ExprKind::Path(_, path) = &base.kind {
+            if let [segment] = path.segments.as_slice() {
+                let ident = segment.ident;
```

Comment:
This doesn't work for code like `foo.bar++`. Instead, we fall back to the base case:

```
error: Rust has no postfix increment operator
 --> bar.rs:2:12
  |
2 |     foo.bar++;
  |            ^^ not a valid postfix operator
  |
  = help: use `+= 1` instead
```

Is there a way to fix this simply?

Reply:
Just guessing from looking at the AST we generate on Compiler Explorer, I assume you don't just need to peel back N layers of `ExprKind::Field` to get to the innermost `ExprKind::Path` and then continue as normal?

Reply:
Looking at this now, I realize this code isn't the issue (we can just use `span_to_snippet` to avoid the AST manipulation). The issue is that when parsing `++foo.bar`, we need to determine how far ahead to parse. Currently it just parses a single path segment, but that doesn't work for `foo.bar`. Thoughts?